presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+        synchronized (this) {
+            checkState(sinkFactories.contains(factory), "Factory is already closed");
+
+            if (!allSourcesFinished) {
+                sinks.add(sink);
+                return sink;
+            }
+        }
+
+        // all sources have already finished so return a finished sink
+        // NOTE: finish must be called outside of the lock since it calls back
+        sink.finished();
+        return sink;
+    }
+
+    private void sinkFinished(LocalExchangeSink sink)
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            sinks.remove(sink);
+        }
+        checkAllSinksComplete();
+    }
+
+    private void noMoreSinkFactories()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            noMoreSinkFactories = true;
+        }
+        checkAllSinksComplete();
+    }
+
+    private void sinkFactoryClosed(LocalExchangeSinkFactory sinkFactory)
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            sinkFactories.remove(sinkFactory);
+        }
+        checkAllSinksComplete();
+    }
+
+    private void checkAllSinksComplete()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            if (!noMoreSinkFactories || !sinkFactories.isEmpty() || !sinks.isEmpty()) {
+                return;
+            }
+        }
+
+        sources.forEach(LocalExchangeSource::finish);
+        writeWaiter.finishAndNotifyWaiters();
+    }
+
+    private ListenableFuture<?> waitForWriting()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            if (bufferBytes.get() < maxBufferedBytes) {
+                return NOT_BLOCKED;
+            }
+        }
+        return writeWaiter.waitFor();
+    }
+
+    private static void checkNotHoldsLock(Object lock)
+    {
+        checkState(!Thread.holdsLock(lock), "Can not execute this method while holding a lock");
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSource
+    {
+        private final List<Type> types;
+        private final Consumer<LocalExchangeSource> onFinish;
+
+        private final Queue<PageReference> buffer = new ConcurrentLinkedQueue<>();
+        private final AtomicLong bufferedBytes = new AtomicLong();
+
+        private final AsyncWaiter readWaiter = new AsyncWaiter();
+
+        @GuardedBy("this")
+        private boolean finishing;
+
+        public LocalExchangeSource(List<? extends Type> types, Consumer<LocalExchangeSource> onFinish)
+        {
+            this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+            this.onFinish = requireNonNull(onFinish, "onFinish is null");
+        }
+
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        public LocalExchangeBufferInfo getBufferInfo()
+        {
+            return new LocalExchangeBufferInfo(bufferedBytes.get(), buffer.size());
```

Comment:
Add a comment saying that this needs to happen outside the lock and observed valued might not be consistent.